exploration/0474-text-vs-code.md

```diff
@@ -0,0 +1,471 @@
+
+---
+
+Users editing a simple message and who wish to add an `input` or `local` annotiation
+to the message do not wish to reformat the message extensively.
+
+Users who have messages that include leading or trailing whitespace
+want to ensure that this whitespace is included in the translatable
+text portion of the message. Which whitespace characters are displayed at runtime
+should not be surprising.
+
+## Requirements
+
+Common things should be easy, uncommon things should be possible.
+
+Developers and translators should be able to read and write the syntax easily in a text editor.
+
+Translators (and their tools) are not software engineers, so we want our syntax
+to be as simple, robust, and non-fussy as possible.
+Multiple levels of complex nesting should be avoided,
+along with any constructs that require an excessive
+level of precision on the part of non-technical users.
+
+As MessageFormat 2 will be at best a secondary language to all its users,
+it should conform to user expectations and require as little learning as possible.
+
+The syntax should avoid footguns,
+in particular as it's passed through various tools during formatting.
+
+ASCII-compatible syntax. While support for non-ASCII characters for variable names,
+values, literals, options, and the like are important, the syntax itself should
+be restricted to ASCII characters. This allows the message to be parsed
+visually by humans even when embedded in a syntax that requires escaping.
+
+Whitespace is forgiving. We _require_ the minimum amount of whitespace and allow
+users to format or change unimportant whitespace as much as they want.
+This avoids the need for translators or tools to be super pedantic about
+formatting.
+
+## Constraints
+
+Limiting the range of characters that need to be escaped in plain text is important.
+Following past precedent,
+this design doc will only consider encapsulation styles which
+start with `{` and end with `}`.
+
+The current syntax includes some plain-ascii keywords:
+`input`, `local`, `match`, and `when`.
+
+The current syntax and active proposals include some sigil + name combinations,
+such as `:number`, `$var`, `|literal|`, `+bold`, and `@attr`.
+
+The current syntax supports unquoted literal values as operands.
+
+Messages themselves are "simple strings" and must be considered to be a single
+line of text. In many containing formats, newlines will be represented as the local
+equivalent of `\n`.
+
+## Proposed Design
+
+**Start in text mode**
+
+In this option, whitespace is trimmed from `pattern` constructs unless
+the pattern is quoted.
+
+```
+Hello world!
+
+Hello {$user}!
+
+{input $now :datetime dateStyle=long}
+Hello {$user}. Today is {$now}
+
+{local $now = {:systemGetCurrentTime :datetime dateStyle=medium}}
+Hello {$user}. Today is {$now}
+
+{match {$count :number integer=true}}
+{when 0}   Hello {$user}. Today is {$now} and you have no geese.
+{when one} Hello {$user}. Today is {$now} and you have {$count} goose.
```

Comment:
The `{when one} Hello {$user}` part makes me really nervous: it looks like two placeholders on both sides of `Hello`. Furthermore, I think the fact that the space left of `Hello` will be trimmed but the one on the right will not is a footgun. I think this is the main reason why I've been opposed to trimming (although I understand which problems it addresses).

Could we consider using a different set of brackets for statements? For example, if we made `#` special in patterns, too, we could consider something like the following:

```
#[one] Hello {$user}.
```

Does this look like less of a footgun now when it comes to the rules about which space will be trimmed and which one won't?

---

Interestingly, there's something about square brackets that I wouldn't actually mind the following:

```
#[one]Hello {$user}.
```

Reply:
See above 😸 

I kind of like this proposal.

```
#match {$expr} {$expr}
#[key key] Hello {$user}
#[key *  ] {$user} hello
#[*   *  ] { Quoted pattern }
```

Or:

```
#match {$expr} {$expr}#[key key] Hello {$user}#[key *  ] {$user} hello#[*   *  ] { Quoted pattern }
```

Reply:
Past feedback from CLDR in particular has been very critical about reserving additional characters than `\{}` in text mode. However, a two-character sequence like `#[` would probably be much more acceptable, given how rare it is in actual message contents.

If we do have a real concern about an all-code-delimited solution looking like the "statement" parts may look too "placeholder", that should be added to the doc.

In the continuing absence of any exemplars of non-i18n-buggy leading or trailing whitespace, that may end up as a sole reason to prefer quoting entire patterns.

Reply:
One quotes the pattern because one is doing something weird inside the pattern. I'm way _way_ **_way_** more cautious about quoting sub-patterns, because those feel more like I18N bugs to me:

```
This is{| |}{$foo}{| |}an I18N bug waiting to happen.

{This is }{$bar}{ another I18N bug waiting to happen.}

{This} {is} {just} {silly} {😸}
```

Interestingly, this past week I was consulting with some folks and saw a string like "\n\n{$placeholder}", which was being used to make an on-screen list (in a `for` loop). That's not really an I18N bug, although it's not normal either. (And actually what it was was a hardcoded string "\n\n" followed by a `+ someVar` that I made them move to externalized with a formatter...)

As I noted in Seville, there are many special-snowflake cases for exterior whitespace. They are not "compelling" use cases and yes one could code around them. If the proposal is to remove quoted patterns, I suppose I could get behind that...

Reply:
> However, a two-character sequence like `#[` would probably be much more acceptable, given how rare it is in actual message contents.

With the warning that `#` is a comment in some of the container formats (for example Java properties and gettext .po files)

Reply:
LOL `#` is our universal placeholder for "sigil to be named later"... we could use a different sigil. `%` or `@`? 

Or model something like unary operators:

```
??match {$foo}
::[foo] I am a pattern
::[*] Me too

??match {$foo}::[foo] I am a pattern::[*] Me too
```

Reply:
One argument for `{Hello {$user}}` instead of `Hello {$user}`

The MF2 strings will likely end up "sprinkled" in existing resource (storage) formats (.properties, .strings, .xml, .json, etc, even templating languages like soy)

And they will coexist with formats used with other APIs.
```
msg1=Plain text, no API, load and use
msg2=Message %!$d, used by printf-like APIs
msg3=Expires on {0}, used by JDK
msg4=Expires on {expDate, date}, used by MF1
msg5=Expires on {expDate, datetime}, used by MF2
```

Unfortunately every single one of them requires different escaping by the API using them (after loading from storage).
`printf` APIs require escaping the `%` by doubling it (`%%`), `\` is used to escape `'`, `"`, and more
MF1 escapes `'` by doubling it, and escapes `{` by prefixing it with `'`.
MF2 has yet another set of escape rules.
(`I'm {foo}` requires escaping the apostrophe in MF1, but not in MF2)

When translating such a file it is really difficult for a human or tool to figure out what the escaping rules are for each string.

This is all about API level escaping, without bringing in the escaping required by the storage file.
And it is all in the Java realm, I did't bring in other languages / frameworks.

My having MF2 wrap the whole message in `{}` we have a strong signal (although not bullet proof) that this is actually MF2 syntax, and what the escaping rules are.

I agree that this status quo is bad, and would be nice to have messages tagged, but this is the real world.
We had countless problems with this before MF2.

Reply:
@mihnita I agree with that wholly.

Because MF2 is not a resource format, our syntax will be embedded into (and thus require) whatever the host format's escaping regime is. That is, this message:

```
match {$foo}
when * {My pattern}
```

Is actually going to be authored like:

```
my.res=match {$foo}\nwhen * {My pattern}
```

This is our argument against using `"` and `'` in the syntax, for example. Our syntax should be as friendly to host formats as possible (but it still needs to work)

Reply:
> With the warning that # is a comment in some of the container formats (for example Java properties and gettext .po files)

OTOH, it's only a comment when it's the first character on a line. When paired with some identifier, this shouldn't be as big of an issue? E.g.:

```properties
my.res=#match {$foo} #[*] My pattern
```

Reply:
Overall, I see potential in the `#[...]` syntax for _code_ (statements). A lot of languages allows these extra directive for code, often called attributes: C++ uses `[[foo]]`, C# uses `[foo]`, Rust uses `#[foo]`.

Reply:
> Interestingly, this past week I was consulting with some folks and saw a string like "\n\n{$placeholder}", which was being used to make an on-screen list (in a `for` loop). That's not really an I18N bug, although it's not normal either. (And actually what it was was a hardcoded string "\n\n" followed by a `+ someVar` that I made them move to externalized with a formatter...)

This appears to be a non-locale-dependent use case of leading whitespace, where the `\n\n` is effectively used as markup, yes? So if (for whatever reason) this message needed to be expressed entirely within MF2, would it make sense to expect this to be represented as `{|\n\n|}`, where the `\n` represent actual newline characters?

One comparable and valid locale-dependent case that I can imagine existing is sentence concatenation in a context that needs to account for both CJK and non-CJK scripts. Similarly to the `\n\n`, I could imagine a space after a period to be included as a leading or trailing space in a single-sentence message for the non-CJK scripts, rather than being handled in code depending on the locale's script.

Are there _**any other**_ locale-dependent uses of leading or trailing whitespace that we ought to consider? And is the case I represent above actual, or purely hypothetical? As in, could someone here state that they do have messages like this in their corpus? _And if they do_, how do they communicate to their translators that in these particular cases, the whitespace should be removed in CJK locales, while something like the `\n\n` above should be left alone?

Reply:
> Are there any other locale-dependent uses of leading or trailing whitespace that we ought to consider? And is the case I represent above actual, or purely hypothetical?

There are lots of locale-dependent cases too. I cited a random example that came my way in the past week because I don't have the luxury of grepping an employer's vast collection of strings at the moment. I'm hopeful someone else can do that, but I don't expect to learn anything from it.

> how do they communicate to their translators that in these particular cases, the whitespace should be removed in CJK locales, while something like the \n\n above should be left alone?

"Carefully." I have seen comments in resource files, comments in translation kits, comments in tooling. I also note that most localization engineering shops maintain tools for checking that target language strings match source strings in terms of start/end spacing, punctuation, and placeables. Some languages (such as CJK) produce a lot of noise or the need for message suppression or tuning in these cases.

I don't have a problem with users putting `{|  \n  \t  |}` onto the front of a pattern as a quoted blob to preserve across translations. But you appear to be building towards the suggestion of *only* allowing that case. I think that format is inconvenient to write and runs up against languages that want other behavior--and also that this isn't really a problem for most translation processes (where the whitespace is already treated as meaningful)

Either way we have to describe how exterior whitespace is handled in patterns. So it doesn't really help me, individually, decide how to choose between auto-trimmed vs. non-trimmed unquoted patterns. I think the thing I'm trying to puzzle out for myself is: which is the most natural and least surprising representation of a pattern?

```
when [*]No whitespace is no whitespace.
when [*] Whitespace is trimmed.
when [*]
    All whitespace is trimmed.
when [*] Whitespace is meaingful, so there is a space before this string.
when [*]\n   All whitespace is meaningful. // read \n as newline
when [*]
   {
   Unquoted whitespace is trimmed, but this message has newline and spaces around it
   }
when [*]{\n   The same message normalized to a single line with\n   }
```

Reply:
> There are lots of locale-dependent cases too. [...] I don't have the luxury of grepping an employer's vast collection of strings at the moment. I'm hopeful someone else can do that, but I don't expect to learn anything from it.

I really hope someone can do that. I have heard now on numerous occasions that there are many locale-dependent uses for leading or trailing whitespace, but this specific one -- how CJK scripts do not use spaces between clauses -- is literally the only one I am aware of.

If we end up making a specific accommodation for this in MF2, this argument needs to be really well made, and presented in this design doc. I am not the right person for doing that, and so I continue to ask others to help here.

> [...] Some languages (such as CJK) produce a lot of noise or the need for message suppression or tuning in these cases.

Could we improve that experience? Rather than perpetuating a sub-optimal CJK translation experience, could we somehow explicitly differentiate localizable leading/trailing whitespace from markup whitespace? If those spaces needed to be explicit, and we did adopt expression attributes, that could be done with

```
{| | @translate}
```

> I don't have a problem with users putting `{| \n \t |}` onto the front of a pattern as a quoted blob to preserve across translations. But you appear to be building towards the suggestion of _only_ allowing that case. I think that format is inconvenient to write and runs up against languages that want other behavior--and also that this isn't really a problem for most translation processes (where the whitespace is already treated as meaningful)

My claim is that almost all leading & trailing whitespace is not really localizable content, and by default should not be in messages. I do still want to allow for the possibility of including such whitespace, and having a way of making it clear to both humans and tooling when such whitespace is markup, and when it is localizable.

We need to identify and enumerate the explicit use cases for leading/trailing whitespace, and to make our syntax choices based on that. Thus far, we have not done so; we've just accepted the assertion that leading/trailing whitespace must be accommodated for ergonomically. Once we do have such a list written down somewhere, we may use that to direct our choice, e.g. within the context of the #485 beauty contest.

And yes, until convinced otherwise, my current position is indeed that we should require leading/trailing whitespace to be explicitly quoted, because that's the only way to differentiate localizable and non-localizable whitespace. To make me change my mind, I can imagine at least the following categories of arguments that could be made:
1. **Locale-dependence**. Show me that there is a great range of different types of [localizable dynamic message strings](https://github.com/unicode-org/message-format-wg/blob/main/docs/goals.md) where leading/trailing whitespace needs to be handled differently in different locales, such that my base assumption about most such cases being "markup" is invalid, and that either both are as common, or that it's more common for there to be a locale dependency.
2. **Better syntax**. Show me syntax which will lead developers to leaving more non-localizable whitespace out of messages and/or communicating better to translators the localizability of whitespace.
3. **Numbers**. Look into the data, i.e. the corpus of localizable dynamic message strings that you have access to, and tell me that such a large fraction of them include valid, localizable leading/trailing whitespace, that special accommodation must be made for them in the syntax. Extra credit if you can further say how common this is in multi-variant messages. I don't need to see your messages, I'm interested in the frequency, and a data-driven argument.

Reply:
> My claim is that almost all leading & trailing whitespace is not really localizable content, and by default should not be in messages. I do still want to allow for the possibility of including such whitespace, and having a way of making it clear to both humans and tooling when such whitespace is markup, and when it is localizable.

I think your claim is reasonable. But I would also observe that this is less true for desktop and CLI applications. I would also note that _trailing_ whitespace is probably as important as leading whitespace.

Overall, I think my reaction on this thread is to look to overall syntax first and whitespace handling as a downstream consideration. A number of syntax options end up with pattern quoting that may make this discussion moot. And, as noted above, each of the whitespace handling options represent different compromises, depending on who is looking and what the use case is.